text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,289 @@
+* Users who would like to make different optimizations to `core` or `std`, such as `opt-level = 'z'`, with `panic = "abort"`
+
+Previously, these needs were somewhat addressed by the external tool [xargo], which managed the recompilation of these dependencies when necessary. However, this tool has become [deprecated], and even when supported, required a nightly version of the compiler for all operation.
+
+This approach has [gathered support] from various [rust team members], and this RFC aims to take inspiration from tools and workflows like [xargo], and integrate them into Cargo itself.
+
+[xargo]: https://github.com/japaric/xargo
+[deprecated]: https://github.com/japaric/xargo/issues/193
+[gathered support]: https://github.com/japaric/xargo/issues/193#issuecomment-359180429
+[rust team members]: https://www.ncameron.org/blog/cargos-next-few-years/
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This proposal aims to make `core` and `std` feel a little bit less like a special case compared to other dependencies to the end users of Cargo. This proposal aims to minimize the number of new concepts introduced to achieve this, by interacting, configuring, modifying, and patching `core` and `std` in a similar manner to other dependent crates.
+
+This RFC proposes the following concrete changes, which may or may not be implemented in this order, and may be done incrementally. The details and caveats around these stages are discussed in the Reference Level Explanation.
+
+1. Allow developers of root crates to recompile `core` (and `compiler-builtins`) when their desired target does not match one available as a `rustup target add` target, without the usage of a nightly compiler. This version of `core` would be built from the same source files used to build the current version of `rustc`/`cargo`.
+2. Introduce the concept of "stable features" for `core`, which allow the end user to influence the behavior of their custom version of `core`, without the use of a nightly compiler.
+3. Extend the new behaviors described in step 1 and 2 for `std` (and `alloc`).
+4. Allow the user to provide their own custom source versions of `core` and `std`, allowing for deep customizations when necessary. This will require a nightly version of the compiler.
+
+As a new concept, the items above propose the existence of "stable features" for `core` and `std`. These features would be considered stable with the same degree of guarantees made for stability in the rest of the language. These features would allow configuration of certain functionalities of `core` or `std`, in a way decided at compile time.
+
+For example, we could propose a feature called `force-tiny-fmt`, which would use different algorithms to implement `fmt` for use on resource constrained systems. The developer of the root crate would be able to choose the default behavior, or the `force-tiny-fmt` behavior while still retaining the ability of using a stable compiler.
```

Comment:
You may want to consider interactions with https://github.com/rust-lang/rfcs/pull/2492 ? (or if not relevant, explain why not somewhere...)

Reply:
I am unsure how this is applicable, was that the right link? If so, could you please expand on what you mean? (The link currently goes to existential types).

The `force-tiny-fmt` feature flag I describe is purely theoretical, and is only used as an example for this RFC. If I need to make that more clear, please let me know.

Reply:
(the link was intended)

So the way it might work is akin to `#[global_allocator]` and `#[panic_handler]` in that the standard library defines an `extern existential type DebugImplementation: SomeTrait = TheDefaultOne;` and then you can override that.

Reply:
Ah, interesting! I was not aware of how these worked under the hood. Could you suggest any possible items to discuss here? Maybe as an addition to the open questions section?

I am proposing this as mostly just a consumer of `core`/`std`/`cargo`, so I am not fully aware of the full implications here (outside of discussions with some of the core team).

Reply:
For now I'll cc @Ericson2314 since they wrote the RFC :)

Reply:
As long as initially the features themselves are unstable (even if the mechanism for stable features exists), I wouldn't worry about that just yet :).